### ISSUE TITLE
(Bug report) Invalid CSRF token after fresh docker installation

### ISSUE THREAD

**Issue by fengwang** [zadam/trilium#4186]
### Trilium Version

0.60.4

### What operating system are you using?

Other Linux

### What is your setup?

Server access only

### Operating System Version

Archlinux

### Description

1. Create an docker container: `mkdir trilium-data && chmod 777 trilium-data && docker run -d --name trilium  -p 0.0.0.0:7582:8080 -v /data/Dockers/Trilium/trilium-data:/home/node/trilium-data --restart unless-stopped zadam/trilium:0.60.4`
2. Visit the web interface, setup as a new user with password and login
3. The web interface shows gray background with message `403 POST tree/load` and `Invalid CSRF token`

Full log is attached within this issue.

[fulllog.txt](https://github.com/zadam/trilium/files/12384672/fulllog.txt)

### Error logs

The docker logs gives error message such as

```
ERROR: Invalid CSRF token: gTy0zmzL-zDpCkMplqlGn6EfAwLnopIyKyLA, secret: zpN74Smky9YA4e1Okk1Ky3hNfqvtHPkk
Error: Invalid CSRF token
    at /usr/src/app/src/app.js:98:11
    at Layer.handle_error (/usr/src/app/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/usr/src/app/node_modules/express/lib/router/index.js:326:13)
    at /usr/src/app/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/usr/src/app/node_modules/express/lib/router/index.js:346:12)
    at next (/usr/src/app/node_modules/express/lib/router/index.js:280:10)
    at /usr/src/app/node_modules/express/lib/router/index.js:646:15
    at next (/usr/src/app/node_modules/express/lib/router/index.js:265:14)
    at next (/usr/src/app/node_modules/express/lib/router/route.js:136:14)
    at Layer.handle_error (/usr/src/app/node_modules/express/lib/router/layer.js:67:12) {
  status: 403
}
JS Error: 403 POST tree/load - Invalid CSRF token
Stack: Error
    at Object.N [as logError] (http://10.147.19.201:7582/assets/v0.60.4/app-dist/desktop.js:2:103948)
    at Object.throwError (http://10.147.19.201:7582/assets/v0.60.4/app-dist/desktop.js:2:89111)
    at h (http://10.147.19.201:7582/assets/v0.60.4/app-dist/desktop.js:2:85051)
    at async Object.error (http://10.147.19.201:7582/assets/v0.60.4/app-dist/desktop.js:2:85367)
Generated CSRF token Y9LGRnOE-gQav_WJtdD7P_Rj0okHpm9rMUas with secret undefined
```

**Comment by fengwang** [comment]
Please note that I encountered this or same problem with many versions, from 0.58.6 to latest

**Comment by zadam** [comment]
I tested the exact command on Ubuntu and it works fine. I wonder what's the differentiating factor here, Arch or perhaps browser, some browser (privacy) extension? I assume there isn't any reverse proxy at play, but perhaps something else configured in the network stack?

**Comment by fengwang** [comment]
Hi @zadam, many thanks for looking into this issue. I created the docker container in a server in my Lan, and visit it remotely from my laptop using a zerotier network. Could this setup be a problem?

**Comment by YodaDaCoda** [comment]
I often get this when going back to a previously-opened trilium tab on Firefox on Android. Seems the tab process ends, and is restored / the page reloaded upon going back to the tab. Unsure of if the timing involved, in my encounters it's in the realm of hours. I'll try to get logs next time it happens.

**Comment by briped** [comment]
Getting the same error, both using the desktop app or the web interface. The only way I've found to get access again, is to clear the site cookies.

I haven't sat down and found a consistent way to reproduce the issue. Right now it seems to happen when I've been updating notes, then shutting down the computer for the day, and the next day the issue often appears.

Running TrilliumNext (same issue running Trillium) in docker, behind a Traefik proxy.

```yaml
networks:
  proxy:
    external: true

services:
  triliumNext:
    image: triliumnext/notes:latest
    container_name: triliumNext
    hostname: real-hostname
    restart: unless-stopped
    environment:
      - TZ=UTC
      - USER_UID=12345
      - USER_GID=12345
    volumes:
      - /home/user/.local/share/triliumNext-data:/home/node/trilium-data
    networks:
      proxy:
    labels:
      - traefik.enable=true
      - traefik.docker.network=proxy
      - traefik.http.routers.rtr-triliumNext.entryPoints=https
      - traefik.http.routers.rtr-triliumNext.rule=Host(`sub.domain.tld`) && ClientIP(`192.0.2.0/24`)
      - traefik.http.services.svc-triliumNext.loadBalancer.server.scheme=http
      - traefik.http.services.svc-triliumNext.loadBalancer.server.port=8080
      - traefik.http.routers.rtr-triliumNext.service=svc-triliumNext
```

**Comment by roughnecks** [comment]
Hello,
I'm running Trilium in docker, listening on 127.0.0.1 and using nginx as a reverse proxy to 192.168.1.125:port, HTTP

Since yesterday I'm getting the same message as OP: Invalid CSRF token
I know this is an old issue, but it's still open, so..

<img width="725" height="397" alt="Image" src="https://github.com/user-attachments/assets/52d7db14-bf33-49ef-8773-26ae10193113" />

Do you have any advice? It's getting frustrating.
Thanks

**Comment by sleak-lbl** [comment]
I'll add a "me too", but I'm just running TriliumNext in a docker container, and accessing it via http://localhost:port. No reverse proxy or anything

<img width="1342" height="638" alt="Image" src="https://github.com/user-attachments/assets/a018293b-bd8a-4577-b0cc-369d8cbc8009" />

**Comment by sleak-lbl** [comment]
Update: it looks like pointing my browser at 127.0.0.1:port instead of localhost:port might resolve it, since experimenting with that change yesterday, I haven't seen any CSRF error messages

**Comment by sleak-lbl** [comment]
This started happening again recently after a firefox update, and after some digging the culprit seems to be firefox's "enhanced tracking protection" - adding an exception for "127.0.0.1" seems to have fixed it

**Comment by fengwang** [comment]
> [@dosu](https://go.dosu.dev/dosubot).

@dosu Please do not dump AI slop here.

**Comment by eliandoran** [comment]
@fengwang , @roughnecks , do you still have issues with CSRF?

**Comment by roughnecks** [comment]
At the moment it's running ok for me. Didn't happen anymore.

**Comment by eliandoran** [comment]
I think it's fine to close it as stale for now, as we've made quite a few changes to the CSRF management since then. If it still occurs we can have new issues with fresh details.

**Comment by KusAdama** [comment]
I'm still experiencing this error.

App version:	0.101.3
DB version:	233
Sync version:	36
Build date:	Jan 8, 2026, 5:45:12 PM
Build revision:	[f5e882271895cca46170dd1bb85be42518fd7841]

Running in Docker and accessing the web UI on local LAN.

If the UI remain opened for a while and the window is not refreshed, I'm getting 403.

Don't now for how long does this happen, because if something doesn't work I just reload with the F5 and continue.
But have feeling it may be month or so.

Trilium logs for a attempt to create new note:
```
ERROR: Invalid CSRF token: <some_token>, secret: <some_secret>
utt [ForbiddenError]: Invalid CSRF token
    at /usr/src/app/main.cjs:2135:89127
JS Error: 403 POST notes/5hxhQiBieGMm/children?target=into&targetBranchId= - Invalid CSRF token
Stack: logError$1@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:228367
reportError@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:5024
async*error@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:3034
f@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:25162
fireWith@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:25922
Se@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:5870
$r/</</send/t/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:8304
EventHandlerNonNull*send@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:8490
ajax@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:5175
ajax/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:3270
ajax@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:2655
call@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:2430
async*post@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:1502
createNote@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:302:50
doRender/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/note_tree.js:116:750
dispatch@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:39612
$r/</add/H.handle@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:37586


JS Error: Uncaught error: Message: undefined, Line: undefined, Column: undefined,
Error object: "{\"message\":\"Invalid CSRF token\"}", Stack: undefined
Stack: logError$1@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:228367
Ke/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/index_jquery.js:1:1383
```

**Comment by eliandoran** [comment]
@KusAdama , interesting. Do you restart the server by any chance during this period? For example due to updates or a scheduled task. 

And can you describe a bit more how do you maintain the web page opened for a month? Is it like a background tab in the browser, is the device & browser on during the entire period?

**Comment by KusAdama** [comment]
> [@KusAdama](https://github.com/KusAdama) , interesting. Do you restart the server by any chance during this period? For example due to updates or a scheduled task.
> 

`docker compose ps -a` > status Up for 6 days. Past six days happened this to me 2x.

> And can you describe a bit more how do you maintain the web page opened for a month? Is it like a background tab in the browser, is the device & browser on during the entire period?

What I mean by 'it may be month or so.' is this 403 has been occurring for approximately a month.
Remembering that I did stuck for a long time on older version, because of `image: triliumnext/notes:latest` instead  of ` image: triliumnext/trilium:stable` in docker-compose. With that older version, this does not happening, but unfortunately I can't remember the version I stuck on.

I don't have the browser opened that long, usually just days, sometimes a week or so. Device goes to sleep/hibernation.
Page is the pinned one.

Basically how it behave is that upon clicking on new note, just nothing is happening, no popup error falling down from top of the screen (like when Traefik changed parsing of the URL recently). Just no response from the Trilium on '+' button.

Also I noticed (as searching for version info) the settings tab just won't load if Trilium is in the state of 403 and instead the New tab pops out and there is 'Open a note by typing the note's title' and the search box instead of settings.

Meanwhile I did the page reload, just to confirm it helps (so no container restart is needed) and so now can't reproduce manually, until it happen again.

**Comment by KusAdama** [comment]
And now it happen again, so within an hour.

Noticed now, because search doesn't work, (there is no action upon clicking on note in result list).

F5 of Trilium page make it work.

 This is what log look like for a search:

<details>

```
Caught WRONG_FINAL_BLOCK_LENGTH, returning cipherText instead
Caught WRONG_FINAL_BLOCK_LENGTH, returning cipherText instead
Slow 304 GET /api/quick-search/<some_search> with 1648 bytes took 486ms
ERROR: Invalid CSRF token: <some_token>, secret: <some_secret>
utt [ForbiddenError]: Invalid CSRF token
    at /usr/src/app/main.cjs:2135:89127
    at Y8.handleError (/usr/src/app/main.cjs:929:25796)
    at b (/usr/src/app/main.cjs:929:31162)
    at /usr/src/app/main.cjs:929:30782
    at qHi (/usr/src/app/main.cjs:929:32791)
    at g (/usr/src/app/main.cjs:929:30714)
    at /usr/src/app/main.cjs:929:33594
    at g (/usr/src/app/main.cjs:929:30633)
    at c (/usr/src/app/main.cjs:929:27871)
    at Y8.handleError (/usr/src/app/main.cjs:929:25781) {
  statusCode: 403
}
JS Error: 403 POST tree/load - Invalid CSRF token
Stack: logError$1@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:228367
reportError@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:5024
async*error@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:3034
f@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:25162
fireWith@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:25922
Se@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:5870
$r/</</send/t/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:8304
EventHandlerNonNull*send@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:8490
ajax@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:4:5175
ajax/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:3270
ajax@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:2655
call@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:2430
async*post@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:1502
reloadNotes@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:23920
getNotes@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:25004
getNote@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:25285
resolveNotePathToSegments@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:11:429
resolveNotePath@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:235374
getResolvedNotePath@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:11:8578
setNote@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:11:7080
displayMoreResults/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/LauncherContainer.js:279:415
dispatch@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:39612
$r/</add/H.handle@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:37586
EventListener.handleEvent*add@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:38052
$r/</vt/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:37223
each@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:2905
each@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:1373
vt@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:37199
on@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:44570
displayMoreResults@http://<a_LAN_ip>:9191/assets/v0.101.3/src/LauncherContainer.js:279:314
search@http://<a_LAN_ip>:9191/assets/v0.101.3/src/LauncherContainer.js:275:570
async*doRender/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/LauncherContainer.js:271:561
dispatch@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:39612
$r/</add/H.handle@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:37586
trigger@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:2:1067
$r/</trigger/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:2:1669
each@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:2905
each@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:1:1373
trigger@http://<a_LAN_ip>:9191/assets/v0.101.3/src/jquery.js:2:1645
trigger@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:140548
show@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:163337
doRender/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/LauncherContainer.js:271:948
Z@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:103968


JS Error: Uncaught error: Message: undefined, Line: undefined, Column: undefined, Error object: "{\"message\":\"Invalid CSRF token\"}", Stack: undefined
Stack: logError$1@http://<a_LAN_ip>:9191/assets/v0.101.3/src/app_context.js:9:228367
Ke/<@http://<a_LAN_ip>:9191/assets/v0.101.3/src/index_jquery.js:1:1383
```
</details>

**Comment by KusAdama** [comment]
Now on 0.102.0 and it's not happening so far.

**Comment by eliandoran** [comment]
@KusAdama , sounds good.
The next version should have an even better CSRF handling.

**Comment by fengwang** [comment]
I confirm this bug is gone with image `triliumnext/trilium:sha-ea20611`, which was released one hour ago.
